### PR TITLE
BUG: special.logsumexp: fix precision issue

### LINKED ISSUE #18295: BUG: special: Loss of precision in `logsumexp`
### Describe your issue.

`logsumexp` loses precision in cases where it shouldn't, and in extreme cases it can return 0 when it should return a nonzero float.

### Reproducing Code Example

```python
In [2]: from scipy.special import logsumexp

In [3]: logsumexp([0, -40.0])
Out[3]: 0.0

In [4]: np.logaddexp(0, -40)  # This is the expected result of `logsumexp([0, -40.0])`.
Out[4]: 4.248354255291589e-18
```

<details>
<summary>

### SciPy/NumPy/Python version and system information

</summary>

```shell
{
  "Compilers": {
    "c": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "11.3.0",
      "commands": "cc"
    },
    "cython": {
      "name": "cython",
      "linker": "cython",
      "version": "0.29.33",
      "commands": "cython"
    },
    "c++": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "11.3.0",
      "commands": "c++"
    },
    "fortran": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "11.3.0",
      "commands": "gfortran"
    },
    "pythran": {
      "version": "0.12.1",
      "include directory": "../../../../../py3.10.8/lib/python3.10/site-packages/pythran"
    }
  },
  "Machine Information": {
    "host": {
      "cpu": "x86_64",
      "family": "x86_64",
      "endian": "little",
      "system": "linux"
    },
    "build": {
      "cpu": "x86_64",
      "family": "x86_64",
      "endian": "little",
      "system": "linux"
    },
    "cross-compiled": false
  },
  "Build Dependencies": {
    "blas": {
      "name": "openblas",
      "found": true,
      "version": "0.3.20",
      "detection method": "pkgconfig",
      "include directory": "/usr/include/x86_64-linux-gnu/openblas-pthread/",
      "lib directory": "/usr/lib/x86_64-linux-gnu/openblas-pthread/",
      "openblas configuration": "USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER=1 NO_CBLAS= NO_LAPACK= NO_LAPACKE=1 NO_AFFINITY=1 USE_OPENMP=0 generic MAX_THREADS=64",
      "pc file directory": "/usr/lib/x86_64-linux-gnu/pkgconfig"
    },
    "lapack": {
      "name": "openblas",
      "found": true,
      "version": "0.3.20",
      "detection method": "pkgconfig",
      "include directory": "/usr/include/x86_64-linux-gnu/openblas-pthread/",
      "lib directory": "/usr/lib/x86_64-linux-gnu/openblas-pthread/",
      "openblas configuration": "USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER=1 NO_CBLAS= NO_LAPACK= NO_LAPACKE=1 NO_AFFINITY=1 USE_OPENMP=0 generic MAX_THREADS=64",
      "pc file directory": "/usr/lib/x86_64-linux-gnu/pkgconfig"
    }
  },
  "Python Information": {
    "path": "/home/warren/py3.10.8/bin/python3",
    "version": "3.10"
  }
}
```

</details>

### PR DESCRIPTION
**Pull request by sadigulcelik**
Closes #18295. Updates scipy.special.logsumexp calculation with more precise logsumexp calculation outlined in #18295 where possible.

### CODE REVIEW

File: scipy/special/tests/test_logsumexp.py [hunk 1]
```diff
@@ -139,6 +139,11 @@ def test_logsumexp_b_shape():
 
     logsumexp(a, b=b)
 
+def test_logsumexp_precision():
+    a = [0, -40]
+    expected = 4.248354255291589e-18
+    assert_almost_equal(logsumexp(a), expected)
```
Comment (MatteoRaso) [inline scipy/special/tests/test_logsumexp.py:145]:
```suggestion
    assert_allclose(logsumexp(a), expected, rtol=1e-18)
```
`assert_almost_equal` doesn't throw an error for precision this low, which might cause a false negative.

**Review by MatteoRaso — CHANGES_REQUESTED** [review]

File: scipy/special/_logsumexp.py [hunk 2]
```diff
@@ -104,19 +104,43 @@ def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
...
         a_max = 0
 
+    tmp0 = (a == a_max) # compute largest separately
+
+
     if b is not None:
         b = np.asarray(b)
-        tmp = b * np.exp(a - a_max)
+        tmp = b * np.exp(a - a_max)  * (~tmp0)
+        tmp0 = b*(tmp0 + .0)
```
Comment (ev-br) [inline scipy/special/_logsumexp.py:113]: This is to convert a bool array to float? Prefer more explicit `tmp0 = (a == a_max).astype(float)` or some such. 

Also might want to make it a bit more explicit with boolean indexing, like `mask = a == a_max`, `a1 = a[~mask]` etc.

File: scipy/special/_logsumexp.py [hunk 3]
```diff
@@ -104,19 +104,43 @@ def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
...
     else:
-        tmp = np.exp(a - a_max)
+        tmp = np.exp(a - a_max)  * (~tmp0)
```
Comment (ev-br) [inline scipy/special/_logsumexp.py:115]: So this pulls out the values where a == a_max to be able to use log1p below. Needs a comment (something along the lines of what Warren wrote in the OP). One option is to write a formula in a comment.

File: scipy/special/_logsumexp.py [hunk 4]
```diff
@@ -104,19 +104,43 @@ def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
...
+        sf = s + s0
         if return_sign:
-            sgn = np.sign(s)
+            sgn = np.sign(sf)
             s *= sgn  # /= makes more sense but we need zero -> zero
-        out = np.log(s)
+            s0*=sgn
+            sf*=sgn
+
+        if(np.isscalar(s0)):
```
Comment (ev-br) [inline scipy/special/_logsumexp.py:129]: Best avoid using `np.isscalar`.  Not sure if it can be a scalar here at all, this comes from `np.sum(..., keepdims=True)`, so it's a 0D array at most.

File: scipy/special/_logsumexp.py [hunk 5]
```diff
@@ -104,19 +104,43 @@ def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
...
+            if(s0<=0):
```
Comment (ev-br) [inline scipy/special/_logsumexp.py:130]: Nit: `if s <= 0:`  (extra brackets, <space> binop <space>). Here and elsewhere

File: scipy/special/tests/test_logsumexp.py [hunk 6]
```diff
@@ -139,6 +139,11 @@ def test_logsumexp_b_shape():
 
     logsumexp(a, b=b)
 
+def test_logsumexp_precision():
+    a = [0, -40]
+    expected = 4.248354255291589e-18
```
Comment (ev-br) [inline scipy/special/tests/test_logsumexp.py:144]: Please add a comment on how to reproduce this value. mpmath or Wolfram Mathematica incantation or how whatever other way you used to reproduce this.

**Review by ev-br — CHANGES_REQUESTED** [review]
Looks like a good start!

I think the implementation could use some explanatory comments and readability tweaks.
Also would be good to add a couple of tests of a similar accuracy problems with higher dimensional arrays and exercise non-default axis and/or keepdims.

File: scipy/special/_logsumexp.py [hunk 7]
```diff
@@ -104,19 +104,60 @@ def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
...
     else:
-        tmp = np.exp(a - a_max)
+        tmp = np.exp(a - a_max)  * (~tmp0)
+        tmp0 = tmp0.astype(float)
 
     # suppress warnings about log of zero
     with np.errstate(divide='ignore'):
         s = np.sum(tmp, axis=axis, keepdims=keepdims)
+        s0 = np.sum(tmp0, axis=axis, keepdims=keepdims)
+        sf = s + s0
```
Comment (lorentzenchr) [inline scipy/special/_logsumexp.py:133]:
```suggestion
        s = s0 + s1
```
It would be nice if the result, ie the sum of both terms, is called s as before.

File: scipy/special/_logsumexp.py [hunk 8]
```diff
@@ -104,19 +104,60 @@ def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
...
         if return_sign:
-            sgn = np.sign(s)
+            sgn = np.sign(sf)
             s *= sgn  # /= makes more sense but we need zero -> zero
-        out = np.log(s)
+            s0*=sgn
+            sf*=sgn
+
+        
+        precise = ((s0>0) & (s>0))
```
Comment (lorentzenchr) [inline scipy/special/_logsumexp.py:141]: From here on, it seems to me that the code could be simpler. s0 and s are either 0 or positive and s==0 does not hurt. So, only special casing for s0==0 should be sufficient.

**Review by lorentzenchr — APPROVED** [review]
Overall, it looks good. Some improvements might be possible.

**Comment by dschmitz89** [comment]
Not trying to block this PR but wouldn't it make sense in the long run to implement `logsumexp` in cython or C so that it can be exposed via `cython_special` too? Sklearn carries a cython implementation, another case of duplication across the ecosystem.

**Comment by lorentzenchr** [comment]
> Not trying to block this PR but wouldn't it make sense in the long run to implement `logsumexp` in cython or C so that it can be exposed via `cython_special` too? Sklearn carries a cython implementation, another case of duplication across the ecosystem.

Note that even if it were available in scipy Cython API, scikit-learn would probably still have its own Cython implementation to make inlining possible. Same story is already true for `expit`. This makes some performance difference that I measured once.

**Comment by lucascolley** [comment]
Hey @sadigulcelik , would you like to return to this? There are some comments above to address but it sounds like this was close.

**Comment by ev-br** [comment]
Needs a rebase now. Would be great to decide on this one either way.

**Comment by tylerjereddy** [comment]
This PR hasn't seen commit activity for more than a year, so I think bumping the milestone is the right call. May need a new champion if the original author is busy.

**Comment by lucascolley** [comment]
It looks like this PR will be superseded by gh-21597. Thanks for this anyway @sadigulcelik!
